File: arrays_and_strings/13_roman_to_integer/python/roman_to_int.py
roman = {
    "I" : 1,
    "V" : 5,
    "X" : 10,
    "L" : 50,
    "C" : 100,
    "D" : 500,
    "M" : 1000
}

#Research why I can't do add operation in the if statement block
class Solution:
    def romanToInt(self, s: str) -> int:
        result = 0
        n = len(s)
        for i in range(n):
            if i+1 < n and roman[s[i]] < roman[s[i+1]]:
                # print("minus",s[i])
                result -= roman.get(s[i], 0)
            else:
                # print("add",s[i])
                result += roman.get(s[i], 0)
        return result
    
if __name__ == "__main__":
    test_cases = ["III", "IV", "IX", "LVIII", "MCMXCIV", "I", "MMMCMXCIX"]

    for s in test_cases:
        print(f"Input:       {s!r}")
        print(f"Output:      {Solution().romanToInt(s)}")
        print("-" * 35)

File: arrays_and_strings/13_roman_to_integer/TS/romanToInt.ts
export {};

const ROMAN = {
    "I" : 1,
    "V" : 5,
    "X" : 10,
    "L" : 50,
    "C" : 100,
    "D" : 500,
    "M" : 1000
} as const

type RomanKey = keyof typeof ROMAN;

function romanToInt(roman: string): number{
    let result = 0;
    let n = roman.length;

    for(let i = 0; i < n; i++){
        let current: RomanKey = roman[i] as RomanKey;
        let next: RomanKey = roman[i+1] as RomanKey;
        if (ROMAN[current] < ROMAN[next]){
            result -= ROMAN[current]
        } else{
            result += ROMAN[current]
        }
    }

    return result
}

const testCases: string[] = ["III", "IV", "IX", "LVIII", "MCMXCIV", "I", "MMMCMXCIX"];

for (const s of testCases) {
    console.log(`Input:       ${JSON.stringify(s)}`);
    console.log(`Output:      ${romanToInt(s)}`);
    console.log("-".repeat(35));
}